File: artifacts/api-server/src/lib/trainingManifest.ts
/**
 * CA2 training-dataset manifest verification (Wave Q — Model Discovery, PR-63).
 *
 * `services/composers-assistant-train/build_dataset.py` writes a manifest for
 * every dataset it builds: digests, counts, the split rule and per-example
 * provenance — never the examples. This module is the platform's reading of
 * that object. It re-derives what it can (the split of every work, the split
 * digest, the leak check, the token ceilings) and hands the provenance to the
 * dataset rights proof, so the training run gates on the same proof the
 * arranger pipeline gates on. A manifest that fails any check is refused with
 * every problem named; a manifest that passes yields a verified proof the
 * trainer must carry.
 */
import { createHash } from "node:crypto";
import {
  buildDatasetRightsProof,
  isProofRefusal,
  type DatasetRightsProof,
  type ExampleProvenance,
  type RightsBasis,
} from "./datasetRightsProof";

export const TRAINING_MANIFEST_VERSION = "CA2_TRAINING_MANIFEST_v1";
export const TRAINING_RIGHTS_BASIS_VERSION = "TRAINING_RIGHTS_BASIS_v1";
export const CA2_MAX_LEN = 1650;
export const CA2_VOCAB_SIZE = 1944;
/** `CA2_UNJOINED_v2.1.0_<vocab sha256 prefix>_1944`, as ca2_pins.tokenizer_version() writes it. */
export const CA2_TOKENIZER_VERSION_PATTERN = /^CA2_UNJOINED_v2\.1\.0_[0-9a-f]{8}_1944$/;
export const SPLIT_RULE = "work-level: int(sha256(workId).hex[:8], 16) % 100 -> <90 train, <95 val, else test";

export type TrainingSplit = "train" | "val" | "test";

export type TrainingExampleProvenance = {
  workId: string;
  split: TrainingSplit;
  index: number;
  exampleId: string;
  targetInst: number;
  inputTokens: number;
  labelTokens: number;
};

export type TrainingManifest = {
  version: string;
  builtAt: string;
  task: string;
  tokenizerVersion: string;
  ca2: Record<string, unknown>;
  datasetDigest: string;
  examplesDigest: string;
  splitDigest: string;
  rightsBasis: {
    version: string;
    basisDigest: string;
    datasetDigest: string;
    rightsDigest: string;
    source: { recordId: string; doi: string; subset: string };
    admittedWorkCount: number;
  };
  splitRule: string;
  maxLen: number;
  maxInputTokensSeen: number;
  maxLabelTokensSeen: number;
  counts: {
    examples: number;
    works: number;
    perSplit: Record<TrainingSplit, { examples: number; works: number; inputTokens: number; labelTokens: number }>;
    perTargetInstrument: Record<string, number>;
    perVariant: Record<string, number>;
  };
  workLevelSplitLeak: boolean;
  provenance: TrainingExampleProvenance[];
};

/** The git-ignored file `scripts/export-training-rights-basis.mjs` writes for the Python builder. */
export type ExportedRightsBasis = {
  version: string;
  exportedAt: string;
  source: { recordId: string; doi: string; subset: string };
  datasetDigest: string;
  rightsDigest: string;
  ourAdmittedCount: number;
  authorsAdmittedCount: number;
  /** Admitted by BOTH gates — the only ids a dataset may be built from. */
  admittedWorkIds: string[];
  /** Ours-only and authors-only, so both original sets can be rebuilt for the proof. */
  ourOnlyWorkIds: string[];
  authorsOnlyWorkIds: string[];
  basisDigest: string;
};

const sha256 = (text: string): string => createHash("sha256").update(text).digest("hex");
const HEX64 = /^[0-9a-f]{64}$/;

/** Identical to extract-arranger-tasks.mjs and training_manifest.split_for_work. */
export function splitForWork(workId: string): TrainingSplit {
  const h = parseInt(sha256(workId).slice(0, 8), 16) % 100;
  return h < 90 ? "train" : h < 95 ? "val" : "test";
}

export function splitDigest(provenance: readonly { workId: string; split: string }[]): string {
  const lines = [...new Set(provenance.map((p) => `${p.workId}:${p.split}`))].sort();
  return sha256(lines.join("\n"));
}

/** Identical to training_manifest.rights_basis_digest. */
export function rightsBasisDigest(basis: Omit<ExportedRightsBasis, "basisDigest">): string {
  const ids = [...basis.admittedWorkIds].sort();
  const head = `${TRAINING_RIGHTS_BASIS_VERSION}\n${basis.datasetDigest}\n${basis.rightsDigest}\n${basis.source.recordId}:${basis.source.subset}\n`;
  return sha256(head + ids.join("\n"));
}

export function exportRightsBasis(
  basis: RightsBasis,
  exportedAt: string,
): ExportedRightsBasis {
  const admitted: string[] = [];
  const ourOnly: string[] = [];
  const authorsOnly: string[] = [];
  for (const id of basis.ourAdmitted) (basis.authorsAdmitted.has(id) ? admitted : ourOnly).push(id);
  for (const id of basis.authorsAdmitted) if (!basis.ourAdmitted.has(id)) authorsOnly.push(id);
  admitted.sort();
  ourOnly.sort();
  authorsOnly.sort();
  const body = {
    version: TRAINING_RIGHTS_BASIS_VERSION,
    exportedAt,
    source: basis.source,
    datasetDigest: basis.datasetDigest,
    rightsDigest: basis.rightsDigest,
    ourAdmittedCount: basis.ourAdmitted.size,
    authorsAdmittedCount: basis.authorsAdmitted.size,
    admittedWorkIds: admitted,
    ourOnlyWorkIds: ourOnly,
    authorsOnlyWorkIds: authorsOnly,
  };
  return { ...body, basisDigest: rightsBasisDigest(body) };
}

/** Rebuild the two gate sets from an export, refusing a tampered file. */
export function rightsBasisFromExport(exported: ExportedRightsBasis): RightsBasis {
  if (exported.version !== TRAINING_RIGHTS_BASIS_VERSION) {
    throw new Error(`rights basis version ${exported.version} is not ${TRAINING_RIGHTS_BASIS_VERSION}`);
  }
  if (rightsBasisDigest(exported) !== exported.basisDigest) {
    throw new Error("rights basis digest mismatch: the exported admitted set was altered; refusing");
  }
  return {
    ourAdmitted: new Set([...exported.admittedWorkIds, ...exported.ourOnlyWorkIds]),
    authorsAdmitted: new Set([...exported.admittedWorkIds, ...exported.authorsOnlyWorkIds]),
    datasetDigest: exported.datasetDigest,
    rightsDigest: exported.rightsDigest,
    source: exported.source,
  };
}

export type ManifestVerification =
  | { ok: true; problems: []; proof: DatasetRightsProof; summary: ManifestSummary }
  | { ok: false; problems: string[]; proof: DatasetRightsProof | null; summary: ManifestSummary | null };

export type ManifestSummary = {
  datasetDigest: string;
  splitDigest: string;
  tokenizerVersion: string;
  examples: number;
  works: number;
  perSplit: TrainingManifest["counts"]["perSplit"];
  perTargetInstrument: Record<string, number>;
};

/**
 * Verify a manifest against a rights basis. Every problem is collected, not
 * just the first: a builder that produced three defects should hear three.
 */
export function verifyTrainingManifest(
  manifest: TrainingManifest,
  basis: RightsBasis,
): ManifestVerification {
  const problems: string[] = [];
  if (manifest.version !== TRAINING_MANIFEST_VERSION) problems.push(`version ${manifest.version} is not ${TRAINING_MANIFEST_VERSION}`);
  if (manifest.task !== "infill") problems.push(`task ${manifest.task} is not CA2's fine-tune task "infill"`);
  if (!CA2_TOKENIZER_VERSION_PATTERN.test(manifest.tokenizerVersion ?? "")) {
    problems.push(`tokenizerVersion ${manifest.tokenizerVersion} is not a CA2 unjoined ${CA2_VOCAB_SIZE}-token vocabulary version`);
  }
  for (const [name, value] of [
    ["datasetDigest", manifest.datasetDigest],
    ["examplesDigest", manifest.examplesDigest],
    ["splitDigest", manifest.splitDigest],
  ] as const) {
    if (!HEX64.test(value ?? "")) problems.push(`${name} is not a sha256 hex digest`);
  }
  if (manifest.splitRule !== SPLIT_RULE) problems.push("splitRule differs from the platform's work-level rule");
  if (manifest.maxLen !== CA2_MAX_LEN) problems.push(`maxLen ${manifest.maxLen} is not CA2's MAX_LEN ${CA2_MAX_LEN}`);
  if (manifest.rightsBasis?.datasetDigest !== basis.datasetDigest) problems.push("rightsBasis.datasetDigest does not match the acquisition manifest");
  if (manifest.rightsBasis?.rightsDigest !== basis.rightsDigest) problems.push("rightsBasis.rightsDigest does not match the acquisition manifest");

  const provenance = Array.isArray(manifest.provenance) ? manifest.provenance : [];
  if (!provenance.length) problems.push("manifest carries no provenance: an empty dataset cannot be verified");
  if (provenance.length !== manifest.counts?.examples) {
    problems.push(`counts.examples ${manifest.counts?.examples} != provenance length ${provenance.length}`);
  }

  const worksBySplit: Record<TrainingSplit, Set<string>> = { train: new Set(), val: new Set(), test: new Set() };
  const seenIds = new Set<string>();
  let overLength = 0;
  let wrongSplit = 0;
  let maxIn = 0;
  let maxLab = 0;
  for (const p of provenance) {
    if (seenIds.has(p.exampleId)) problems.push(`duplicate exampleId ${p.exampleId}`);
    seenIds.add(p.exampleId);
    if (splitForWork(p.workId) !== p.split) wrongSplit += 1;
    if (p.inputTokens > CA2_MAX_LEN || p.labelTokens > CA2_MAX_LEN) overLength += 1;
    maxIn = Math.max(maxIn, p.inputTokens);
    maxLab = Math.max(maxLab, p.labelTokens);
    worksBySplit[p.split]?.add(p.workId);
  }
  if (wrongSplit) problems.push(`${wrongSplit} example(s) sit in a split the work-level rule does not assign them`);
  if (overLength) problems.push(`${overLength} example(s) exceed MAX_LEN ${CA2_MAX_LEN}`);
  const leak =
    [...worksBySplit.train].some((w) => worksBySplit.val.has(w) || worksBySplit.test.has(w)) ||
    [...worksBySplit.val].some((w) => worksBySplit.test.has(w));
  if (leak) problems.push("work-level split leak: a work appears in two splits");
  if (manifest.workLevelSplitLeak) problems.push("manifest itself reports a split leak");
  if (splitDigest(provenance) !== manifest.splitDigest) problems.push("splitDigest does not match the provenance");
  for (const s of ["train", "val", "test"] as const) {
    const declared = manifest.counts?.perSplit?.[s];
    if (!declared) { problems.push(`counts.perSplit.${s} missing`); continue; }
    const actual = provenance.filter((p) => p.split === s).length;
    if (declared.examples !== actual) problems.push(`counts.perSplit.${s}.examples ${declared.examples} != ${actual}`);
    if (declared.works !== worksBySplit[s].size) problems.push(`counts.perSplit.${s}.works ${declared.works} != ${worksBySplit[s].size}`);
  }
  if (manifest.maxInputTokensSeen !== maxIn) problems.push(`maxInputTokensSeen ${manifest.maxInputTokensSeen} != ${maxIn}`);
  if (manifest.maxLabelTokensSeen !== maxLab) problems.push(`maxLabelTokensSeen ${manifest.maxLabelTokensSeen} != ${maxLab}`);

  // The gate that matters: every example traces to a work admitted by both gates.
  const examples: ExampleProvenance[] = provenance.map((p) => ({ workId: p.workId, shard: p.split, index: p.index }));
  const proofResult = buildDatasetRightsProof(examples, basis, manifest.tokenizerVersion);
  let proof: DatasetRightsProof | null = null;
  if (isProofRefusal(proofResult)) {
    problems.push(`rights proof refused: ${proofResult.reason}`);
  } else {
    proof = proofResult;
    if (!proofResult.verified) {
      problems.push(
        `rights proof failed: ${proofResult.examplesUntraceable} untraceable, ${proofResult.examplesInAuthorsExcluded} in the authors' excluded set`,
      );
    }
  }

  const summary: ManifestSummary | null = manifest.counts
    ? {
        datasetDigest: manifest.datasetDigest,
        splitDigest: manifest.splitDigest,
        tokenizerVersion: manifest.tokenizerVersion,
        examples: manifest.counts.examples,
        works: manifest.counts.works,
        perSplit: manifest.counts.perSplit,
        perTargetInstrument: manifest.counts.perTargetInstrument,
      }
    : null;

  if (problems.length || !proof) return { ok: false, problems, proof, summary };
  return { ok: true, problems: [], proof, summary: summary! };
}

/** The file the trainer requires next to manifest.json before it will start. */
export function rightsProofRecord(verification: ManifestVerification, verifiedAt: string) {
  return {
    version: "CA2_TRAINING_RIGHTS_PROOF_RECORD_v1",
    verifiedAt,
    ok: verification.ok,
    problems: verification.problems,
    datasetDigest: verification.summary?.datasetDigest ?? null,
    proof: verification.proof,
  };
}

File: services/composers-assistant-train/training_manifest.py
"""
Deterministic training-dataset manifest (Wave Q — Model Discovery, PR-63).

The manifest is the only thing about a dataset that is ever committed: digests,
counts, the split rule and per-example provenance (work id, split, index) —
never an example, never a note. `artifacts/api-server/src/lib/trainingManifest.ts`
verifies the same object on the platform side and produces the dataset rights
proof the trainer gates on.

The split rule is the one `scripts/extract-arranger-tasks.mjs` established:
work-level, by sha256 of the work id, 90/5/5. Every example of a work goes to
one split, so no bar of a score is ever on both sides of a validation.
"""
from __future__ import annotations

import hashlib
import json
from typing import Any, Iterable

MANIFEST_VERSION = "CA2_TRAINING_MANIFEST_v1"
RIGHTS_BASIS_VERSION = "TRAINING_RIGHTS_BASIS_v1"
SPLIT_RULE = "work-level: int(sha256(workId).hex[:8], 16) % 100 -> <90 train, <95 val, else test"
MAX_LEN = 1650


def sha256_text(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def split_for_work(work_id: str) -> str:
    h = int(hashlib.sha256(work_id.encode("utf-8")).hexdigest()[:8], 16) % 100
    return "train" if h < 90 else "val" if h < 95 else "test"


def canonical_json(obj: Any) -> str:
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=True)


def examples_digest(examples: Iterable[dict[str, Any]]) -> str:
    """sha256 over the canonical JSON of every example, in order."""
    h = hashlib.sha256()
    for ex in examples:
        h.update(canonical_json(ex).encode("utf-8"))
        h.update(b"\n")
    return h.hexdigest()


def split_digest(provenance: Iterable[dict[str, Any]]) -> str:
    lines = sorted({f"{p['workId']}:{p['split']}" for p in provenance})
    return sha256_text("\n".join(lines))


def rights_basis_digest(basis: dict[str, Any]) -> str:
    """Digest of an exported rights basis — the admitted intersection plus the acquisition digests."""
    ids = sorted(basis["admittedWorkIds"])
    head = f"{RIGHTS_BASIS_VERSION}\n{basis['datasetDigest']}\n{basis['rightsDigest']}\n{basis['source']['recordId']}:{basis['source']['subset']}\n"
    return sha256_text(head + "\n".join(ids))


def load_rights_basis(path: str) -> dict[str, Any]:
    with open(path, "r", encoding="utf-8") as f:
        basis = json.load(f)
    if basis.get("version") != RIGHTS_BASIS_VERSION:
        raise RuntimeError(f"rights basis version {basis.get('version')!r} != {RIGHTS_BASIS_VERSION!r}")
    actual = rights_basis_digest(basis)
    if actual != basis.get("basisDigest"):
        raise RuntimeError("rights basis digest mismatch: the exported admitted set was altered; refusing")
    if not basis["admittedWorkIds"]:
        raise RuntimeError("rights basis admits no work; refusing")
    return basis


class RightsRefusal(RuntimeError):
    pass


def assert_work_admitted(work_id: str, admitted: set[str]) -> None:
    if work_id not in admitted:
        raise RightsRefusal(f"work {work_id} is not in the rights basis intersection; no example may be built from it")


def check_lengths(input_tokens: int, label_tokens: int, max_len: int = MAX_LEN) -> None:
    """CA2's own rule: input <= MAX_LEN, labels (before EOS) <= MAX_LEN - 1."""
    if input_tokens > max_len:
        raise ValueError(f"input {input_tokens} tokens exceeds MAX_LEN {max_len}")
    if label_tokens > max_len:
        raise ValueError(f"labels {label_tokens} tokens (with EOS) exceed MAX_LEN {max_len}")


def build_manifest(
    *,
    examples: list[dict[str, Any]],
    basis: dict[str, Any],
    tokenizer_version: str,
    ca2: dict[str, Any],
    builder: dict[str, Any],
    dedupe: dict[str, Any],
    shards: dict[str, dict[str, Any]],
    built_at: str,
) -> dict[str, Any]:
    """Examples are the full records (inputIds etc.); the manifest keeps only their provenance and counts."""
    admitted = set(basis["admittedWorkIds"])
    provenance = []
    per_split: dict[str, dict[str, Any]] = {s: {"examples": 0, "works": set(), "inputTokens": 0, "labelTokens": 0} for s in ("train", "val", "test")}
    per_inst: dict[str, int] = {}
    per_variant: dict[str, int] = {}
    max_in, max_lab = 0, 0
    for ex in examples:
        assert_work_admitted(ex["workId"], admitted)
        expected = split_for_work(ex["workId"])
        if ex["split"] != expected:
            raise ValueError(f"example {ex['exampleId']} is in split {ex['split']} but the rule says {expected}")
        check_lengths(ex["inputTokens"], ex["labelTokens"])
        s = per_split[ex["split"]]
        s["examples"] += 1
        s["works"].add(ex["workId"])
        s["inputTokens"] += ex["inputTokens"]
        s["labelTokens"] += ex["labelTokens"]
        per_inst[str(ex["targetInst"])] = per_inst.get(str(ex["targetInst"]), 0) + 1
        per_variant[ex["variant"]] = per_variant.get(ex["variant"], 0) + 1
        max_in, max_lab = max(max_in, ex["inputTokens"]), max(max_lab, ex["labelTokens"])
        provenance.append({
            "workId": ex["workId"], "split": ex["split"], "index": ex["index"], "exampleId": ex["exampleId"],
            "targetInst": ex["targetInst"], "inputTokens": ex["inputTokens"], "labelTokens": ex["labelTokens"],
        })
    train_works, val_works, test_works = (per_split[s]["works"] for s in ("train", "val", "test"))
    leak = bool(train_works & val_works or train_works & test_works or val_works & test_works)
    if leak:
        raise ValueError("work-level split leak: a work appears in two splits")
    ex_digest = examples_digest(examples)
    sp_digest = split_digest(provenance)
    basis_digest = basis["basisDigest"]
    dataset_digest = sha256_text(f"{MANIFEST_VERSION}\n{tokenizer_version}\n{ex_digest}\n{sp_digest}\n{basis_digest}")
    return {
        "version": MANIFEST_VERSION,
        "builtAt": built_at,
        "task": "infill",
        "tokenizerVersion": tokenizer_version,
        "ca2": ca2,
        "datasetDigest": dataset_digest,
        "examplesDigest": ex_digest,
        "splitDigest": sp_digest,
        "rightsBasis": {
            "version": basis["version"],
            "basisDigest": basis_digest,
            "datasetDigest": basis["datasetDigest"],
            "rightsDigest": basis["rightsDigest"],
            "source": basis["source"],
            "admittedWorkCount": len(admitted),
        },
        "splitRule": SPLIT_RULE,
        "maxLen": MAX_LEN,
        "maxInputTokensSeen": max_in,
        "maxLabelTokensSeen": max_lab,
        "counts": {
            "examples": len(examples),
            "works": len(train_works | val_works | test_works),
            "perSplit": {
                s: {"examples": v["examples"], "works": len(v["works"]), "inputTokens": v["inputTokens"], "labelTokens": v["labelTokens"]}
                for s, v in per_split.items()
            },
            "perTargetInstrument": dict(sorted(per_inst.items(), key=lambda kv: int(kv[0]))),
            "perVariant": dict(sorted(per_variant.items())),
        },
        "workLevelSplitLeak": leak,
        "dedupe": dedupe,
        "builder": builder,
        "shards": shards,
        "provenance": provenance,
    }


def verify_manifest_against_shards(manifest: dict[str, Any], shard_records: dict[str, list[dict[str, Any]]]) -> None:
    """Re-derive the digests from shard contents; refuse a manifest that does not describe its shards."""
    ordered: list[dict[str, Any]] = []
    for split in ("train", "val", "test"):
        ordered.extend(shard_records.get(split, []))
    if examples_digest(ordered) != manifest["examplesDigest"]:
        raise RuntimeError("examplesDigest mismatch: shards do not match the manifest")
    if split_digest(manifest["provenance"]) != manifest["splitDigest"]:
        raise RuntimeError("splitDigest mismatch")
    if len(ordered) != manifest["counts"]["examples"]:
        raise RuntimeError("example count mismatch")
